File: formatters/chicago.py
"""
Chicago Author-Date formatter.

Layout:
  - Title page: title centered ~1/3 down the page, then author / course / date.
    NO page number on the title page (different_first_page_header_footer=True).
  - Body: double-spaced, 0.5" first-line indent, centered bold headings.
  - References page: "References" centered (not bold), single-spaced entries
    with a blank line between each, hanging indent.
"""

from docx import Document
from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_LINE_SPACING
from docx.shared import Inches, Pt

from models import EssayJSON
from .base import (
    add_body,
    apply_document_defaults,
    add_page_numbers,
    add_references_page,
    centered_line,
    doc_to_bytes,
    set_run_font,
    setup_section,
)


def build(essay: EssayJSON, config: dict) -> bytes:
    doc = Document()
    apply_document_defaults(doc, config)
    section = doc.sections[0]
    setup_section(section, config)

    # Suppress page number on the title page; default header shows it on all other pages
    section.different_first_page_header_footer = True
    add_page_numbers(section, config)

    _title_page(doc, essay, config)
    doc.add_page_break()
    add_body(doc, essay.body_markdown, config, _heading)
    doc.add_page_break()
    add_references_page(doc, essay.references, config)

    return doc_to_bytes(doc)


# ── Title page ───────────────────────────────────────────────────────────────

def _title_page(doc: Document, essay: EssayJSON, config: dict) -> None:
    # Title centered roughly one-third down the page, plain text (not bold)
    title_para = doc.add_paragraph()
    title_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    title_para.paragraph_format.space_before = Inches(3)
    title_para.paragraph_format.space_after = Pt(0)
    title_para.paragraph_format.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    title_para.paragraph_format.first_line_indent = Inches(0)
    run = title_para.add_run(essay.title)
    set_run_font(run, config)

    for text in [
        essay.author_name_placeholder,
        essay.course_placeholder,
        essay.date,
    ]:
        centered_line(doc, text, config)


# ── Headings ─────────────────────────────────────────────────────────────────

def _heading(doc: Document, text: str, level: int, config: dict) -> None:
    """
    Chicago does not rigidly define heading styles for student papers.
    Centered bold for level 1, left-aligned bold for level 2 is widely accepted.
    """
    p = doc.add_paragraph()
    p.paragraph_format.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    p.paragraph_format.space_before = Pt(0)
    p.paragraph_format.space_after = Pt(0)
    p.paragraph_format.first_line_indent = Inches(0)

    if level == 1:
        p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    else:
        p.alignment = WD_ALIGN_PARAGRAPH.LEFT

    run = p.add_run(text)
    run.bold = True
    set_run_font(run, config)

File: src/Formatters/Chicago.php
<?php

namespace App\Formatters;

use App\Models\EssayJSON;
use PhpOffice\PhpWord\Element\Section;
use PhpOffice\PhpWord\PhpWord;
use PhpOffice\PhpWord\Shared\Converter;

/**
 * Chicago 17 formatter.
 *
 * Layout:
 *   - Title page: title centered ~1/3 down the page, then author / course / date.
 *     NO page number on the title page (differentFirstPage = true).
 *   - Body: double-spaced, 0.5" first-line indent, centered bold headings.
 *   - References page: "References" centered (not bold), single-spaced entries
 *     with a blank line between each, hanging indent.
 */
class Chicago
{
    public static function build(EssayJSON $essay, array $config): string
    {
        $phpWord = new PhpWord();
        Base::applyDocumentDefaults($phpWord, $config);

        $section = $phpWord->addSection();
        Base::setupSection($section, $config);

        // Start at 0: title page = 0 (hidden by empty first header), body page 1 = 1
        $section->getStyle()->setPageNumberingStart(0);
        $section->addHeader('first'); // empty — no page number on title page
        Base::addPageNumbers($section, $config); // page number on all other pages

        self::titlePage($section, $essay, $config);
        $section->addPageBreak();
        Base::addBody($section, $essay->bodyMarkdown, $config, [self::class, 'heading']);
        $section->addPageBreak();
        Base::addReferencesPage($section, $essay->references, $config);

        return Base::docToBytes($phpWord);
    }

    // ── Title page ────────────────────────────────────────────────────────────

    private static function titlePage(Section $section, EssayJSON $essay, array $config): void
    {
        // Title centered roughly one-third down the page, plain text (not bold)
        $titleRun = $section->addTextRun([
            'alignment'   => 'center',
            'spaceBefore' => Converter::inchToTwip(3),
            'spaceAfter'  => 0,
            'lineHeight'  => 2.0,
            'indentation' => ['firstLine' => 0],
        ]);
        $titleRun->addText($essay->title, [
            'name' => $config['font_family'],
            'size' => $config['font_size'],
        ]);

        foreach ([
            $essay->authorNamePlaceholder,
            $essay->coursePlaceholder,
            $essay->date,
        ] as $text) {
            Base::centeredLine($section, $text, $config);
        }
    }

    // ── Headings ──────────────────────────────────────────────────────────────

    public static function heading(Section $section, string $text, int $level, array $config): void
    {
        if ($level === 0) {
            Base::addDocumentTitle($section, $text, $config, false); // Chicago: title NOT bold
            return;
        }

        $base = [
            'lineHeight'  => 2.0,
            'spaceBefore' => 0,
            'spaceAfter'  => 0,
            'indentation' => ['firstLine' => 0],
        ];

        // Chicago heading levels:
        // L1 ##:    centered, bold
        // L2 ###:   left, bold
        // L3 ####:  left, italic
        // L4 #####: centered, italic
        // L5 ######: underlined, left
        $specs = [
            1 => ['alignment' => 'center', 'bold' => true,  'italic' => false, 'underline' => false],
            2 => ['alignment' => 'left',   'bold' => true,  'italic' => false, 'underline' => false],
            3 => ['alignment' => 'left',   'bold' => false, 'italic' => true,  'underline' => false],
            4 => ['alignment' => 'center', 'bold' => false, 'italic' => true,  'underline' => false],
            5 => ['alignment' => 'left',   'bold' => false, 'italic' => false, 'underline' => true],
        ];
        $s = $specs[$level] ?? $specs[5];

        $textRun = $section->addTextRun(array_merge($base, ['alignment' => $s['alignment']]));
        $textRun->addText($text, [
            'name'      => $config['font_family'],
            'size'      => $config['font_size'],
            'bold'      => $s['bold'],
            'italic'    => $s['italic'],
            'underline' => $s['underline'] ? 'single' : 'none',
        ]);
    }
}
